File: src/main/java/xyz/oribuin/lilori/database/SQLiteConnector.java
package xyz.oribuin.lilori.database;

import java.io.File;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

/**
 * @author Esophose
 */

public class SQLiteConnector implements DatabaseConnector {

    private final String connectionString;
    private Connection connection;

    public SQLiteConnector(File directory, String dbName) {
        this.connectionString = "jdbc:sqlite" + new File(directory, dbName + ".db").getAbsolutePath();
        try {
            Class.forName("org.sqlite.JDBC"); // Make sure the driver is actually registered
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean isInitialized() {
        return true; // Always available
    }

    @Override
    public void closeConnection() {
        try {
            if (this.connection != null) {
                this.connection.close();
            }
        } catch (SQLException ex) {
            System.err.println("An error occurred closing the SQLite Database: " + ex.getMessage());
        }
    }

    @Override
    public void connect(ConnectionCallback callback) {
        if (this.connection == null) {
            try {
                this.connection = DriverManager.getConnection(this.connectionString);
            } catch (SQLException ex) {
                System.err.println("An error occurred retrieving the SQLite database connection: " + ex.getMessage());
            }
        }

        try {
            callback.accept(this.connection);
        } catch (Exception ex) {
            System.err.println("An error occurred executing an SQLite query: " + ex.getMessage());
            ex.printStackTrace();
        }
    }

    @Override
    public <T> T connect(Class<T> returnType, ConnectionCallbackReturning<T> callback) {
        if (this.connection == null) {
            try {
                this.connection = DriverManager.getConnection(this.connectionString);
            } catch (SQLException ex) {
                System.err.println("An error occurred retrieving the SQLite database connection: " + ex.getMessage());
            }
        }

        try {
            return callback.accept(this.connection);
        } catch (Exception ex) {
            System.err.println("An error occurred executing an SQLite query: " + ex.getMessage());
            ex.printStackTrace();
        }

        return null;
    }
}

File: src/main/kotlin/xyz/oribuin/lilori/database/SQLiteConnector.kt
package xyz.oribuin.lilori.database

import java.io.File
import java.sql.Connection
import java.sql.DriverManager
import java.sql.SQLException

class SQLiteConnector(directory: File) : DatabaseConnector {
    private val connectionString = "jdbc:sqlite:" + directory.absolutePath
    private var connection: Connection? = null

    init {
        try {
            Class.forName("org.sqlite.JDBC")
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
    }

    override fun closeConnection() {
        try {
            if (connection != null) {
                connection?.close()
            }
        } catch (ex: SQLException) {
            println("An error occurred closing the SQLite database connection: " + ex.message)
        }
    }

    override fun connect(callback: (Connection) -> Unit) {
        if (connection == null) {
            try {
                connection = DriverManager.getConnection(connectionString)
            } catch (ex: SQLException) {
                println("An error occurred retrieving the SQLite database connection: " + ex.message)
            }
        }
        try {
            this.connection?.let { callback(it) }
        } catch (ex: Exception) {
            println("An error occurred executing an SQLite query: " + ex.message)
            ex.printStackTrace()
        }
    }
}